File: apps/server/services/project_lock.py
"""Project-level advisory lock for preventing concurrent mutating operations across stacks.

All stack operations that modify state (apply, destroy, refresh, rollback, strip)
should acquire this lock before starting and release it after finishing.
"""
from __future__ import annotations

import time
import uuid
from typing import Any, Optional

from storage import pg

LOCK_PREFIX = "radas.project_lock:"


def _lock_key(project_id: str) -> str:
    return LOCK_PREFIX + project_id


def acquire(project_id: str, *, actor: str, operation: str, run_id: Optional[str] = None) -> dict[str, Any]:
    """Acquire a project-level advisory lock and record the lease.

    Returns a dict with 'ok' and 'lock' if acquired, or 'ok': False with 'lock' if already held.
    """
    lock_key = _lock_key(project_id)
    now = time.time()
    lease_id = str(uuid.uuid4())

    with pg.transaction() as conn:
        # Acquire advisory lock (transaction-level; released on rollback/commit)
        conn.execute("SELECT pg_advisory_xact_lock(hashtextextended(%s, 0))", (lock_key,))

        # Check if there is an existing active lease
        existing = conn.execute(
            "SELECT * FROM project_locks WHERE project_id = %s AND expires_at > %s FOR UPDATE",
            (project_id, now),
        ).fetchone()
        if existing:
            return {"ok": False, "lock": dict(existing)}

        # Insert new lease
        expires_at = now + 3600  # 1 hour timeout
        conn.execute(
            "INSERT INTO project_locks (id, project_id, actor, operation, run_id, acquired_at, expires_at) "
            "VALUES (%s, %s, %s, %s, %s, %s, %s)",
            (lease_id, project_id, actor, operation, run_id, now, expires_at),
        )
        return {"ok": True, "lock": {"id": lease_id, "project_id": project_id, "actor": actor,
                                     "operation": operation, "run_id": run_id,
                                     "acquired_at": now, "expires_at": expires_at}}


def release(project_id: str, *, lock_id: Optional[str] = None, force: bool = False) -> dict[str, Any]:
    """Release the project lock. If force=True, release regardless of lease id."""
    with pg.transaction() as conn:
        # Acquire advisory lock to prevent race
        conn.execute("SELECT pg_advisory_xact_lock(hashtextextended(%s, 0))", (_lock_key(project_id),))
        if lock_id:
            result = conn.execute(
                "DELETE FROM project_locks WHERE project_id = %s AND id = %s RETURNING *",
                (project_id, lock_id),
            )
        else:
            result = conn.execute(
                "DELETE FROM project_locks WHERE project_id = %s RETURNING *",
                (project_id,),
            )
        if result.rowcount == 0 and not force:
            return {"ok": False, "error": "No active lock found"}
        return {"ok": True, "released": result.rowcount > 0, "previous": result.fetchone()}


def get_lock(project_id: str) -> Optional[dict[str, Any]]:
    """Return the active lock for a project, if any."""
    now = time.time()
    row = pg.query_one(
        "SELECT * FROM project_locks WHERE project_id = %s AND expires_at > %s",
        (project_id, now),
    )
    return dict(row) if row else None


def cleanup_expired() -> int:
    """Remove expired locks."""
    now = time.time()
    with pg.transaction() as conn:
        result = conn.execute("DELETE FROM project_locks WHERE expires_at <= %s", (now,))
        return result.rowcount

File: apps/server/lib/radas_ai/project_lock.ex
defmodule RadasAI.ProjectLock do
  @moduledoc """
  Port of `services/project_lock.py` — project-level advisory lock
  (UC373) preventing concurrent mutating operations across stacks.
  Postgres advisory lock + `project_locks` lease table, 1h expiry.
  """

  import RadasAI.DB

  alias Radas.Repo

  @lock_prefix "radas.project_lock:"

  defp lock_key(project_id), do: @lock_prefix <> (project_id || "default")

  @doc "Acquire the project lease: %{\"ok\" => bool, \"lock\" => lease} (Python acquire)."
  @spec acquire(String.t(), keyword()) :: map()
  def acquire(project_id, opts \\ []) do
    actor = Keyword.get(opts, :actor, "")
    operation = Keyword.get(opts, :operation, "")
    run_id = Keyword.get(opts, :run_id)
    now = now()
    lease_id = Ecto.UUID.generate()

    {:ok, result} =
      Repo.transaction(fn ->
        Repo.query!("SELECT pg_advisory_xact_lock(hashtextextended($1, 0))", [lock_key(project_id)])

        existing =
          query_one!(
            "SELECT * FROM project_locks WHERE project_id = $1 AND expires_at > $2 FOR UPDATE",
            [project_id, now]
          )

        if existing do
          %{"ok" => false, "lock" => existing}
        else
          expires_at = now + 3600

          execute!(
            """
            INSERT INTO project_locks (id, project_id, actor, operation, run_id, acquired_at, expires_at)
            VALUES ($1, $2, $3, $4, $5, $6, $7)
            """,
            [lease_id, project_id, actor, operation, run_id, now, expires_at]
          )

          %{
            "ok" => true,
            "lock" => %{
              "id" => lease_id,
              "project_id" => project_id,
              "actor" => actor,
              "operation" => operation,
              "run_id" => run_id,
              "acquired_at" => now,
              "expires_at" => expires_at
            }
          }
        end
      end)

    result
  end

  @doc "Release the project lease (Python release): %{\"ok\", \"released\"?, \"error\"?}."
  @spec release(String.t(), keyword()) :: map()
  def release(project_id, opts \\ []) do
    lock_id = Keyword.get(opts, :lock_id)
    force = Keyword.get(opts, :force, false)

    {:ok, result} =
      Repo.transaction(fn ->
        Repo.query!("SELECT pg_advisory_xact_lock(hashtextextended($1, 0))", [lock_key(project_id)])

        {sql, params} =
          if lock_id do
            {"DELETE FROM project_locks WHERE project_id = $1 AND id = $2 RETURNING *", [project_id, lock_id]}
          else
            {"DELETE FROM project_locks WHERE project_id = $1 RETURNING *", [project_id]}
          end

        rows = query_all!(sql, params)

        if rows == [] and not force do
          %{"ok" => false, "error" => "No active lock found"}
        else
          %{"ok" => true, "released" => rows != [], "previous" => List.first(rows)}
        end
      end)

    result
  end

  @doc "Active lock for a project, if any (Python get_lock)."
  @spec get_lock(String.t()) :: map() | nil
  def get_lock(project_id) do
    query_one!("SELECT * FROM project_locks WHERE project_id = $1 AND expires_at > $2", [
      project_id,
      now()
    ])
  end

  @doc "Remove expired leases; returns the number of rows removed."
  @spec cleanup_expired() :: integer()
  def cleanup_expired do
    execute!("DELETE FROM project_locks WHERE expires_at <= $1", [now()]) || 0
  end
end
